### PR TITLE
refactor: (new) move containerd and runc from manifest.json to components.json. Merge DownloadFiles to Packages

### PR DESCRIPTION
**Pull request by Devinwong**
**What type of PR is this?**

/kind feature

- Move containerd and runc from manifest.json to components.json->Packages.
- Merge former DownloadFiles to Packages, in components.json
- Components.cue defines the schema and components.json hosts the actual components including ContainerImages and Packages.
- Packages and corresponding versions are only defined in components.json now. User can define single or multiple package versions to download, depending on the use case. User can also different version or download URL according to different OS Distro (Ubuntu/Mariner) and releases (18.04/20.04 etc.). A package's downloadURL can contains unresolved variables for the scripts to resolve later, to provide an additional level of flexibility.
- Add shellspec to do unit tests. If there is better framework, I can migrate the test codes to that too.
- Update the component data model to fit the new schema changes.

AB E2E passed. https://msazure.visualstudio.com/CloudNativeCompute/_build/results?buildId=97984006&view=results

**Which issue(s) this PR fixes**:

Fixes #

**Requirements**:

- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [x] includes documentation
- [x] adds unit tests
- [ ] tested upgrade from previous version

**Special notes for your reviewer**:

**Release note**:

```
none
```

### CODE REVIEW

File: parts/linux/cloud-init/artifacts/cse_install.sh
Comment (Devinwong) [inline parts/linux/cloud-init/artifacts/cse_install.sh:45]: This should be the bug for the E2E test failure.
Added 1 unit test to cover this too.

**Review by djsly — APPROVED** [review]

File: parts/linux/cloud-init/artifacts/cse_install.sh
Comment (Devinwong) [inline parts/linux/cloud-init/artifacts/cse_install.sh:48]: This is another bug. Newly added unit test cover this too.

File: parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh
Comment (cameronmeissner) [inline parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh:118]: did you test to make sure specifying an actual version here works via build logs?

Comment (Devinwong) [inline parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh:118]: we were not specifying mobd-containerd version for Mariner. Adding it here.

File: parts/linux/cloud-init/artifacts/components.json
Comment (ganeshkumarashok) [inline parts/linux/cloud-init/artifacts/components.json:414]: What does "cm2" stand for?

Comment (Devinwong) [reply parts/linux/cloud-init/artifacts/components.json:414]: I don't know either. But I got this version from a previous pipeline that we've been using.
`azure-arm:  moby-runc x86_64 1.1.9-5.cm2 mariner-official-base 3.4 M`

File: parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh
Comment (Devinwong) [reply parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh:118]: yes I tested it manually. Let me paste the link.

Comment (Devinwong) [reply parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh:118]: https://msazure.visualstudio.com/CloudNativeCompute/_build/results?buildId=96177599&view=logs&j=da6da462-43f6-5106-fd69-5730003ba730&t=274a1bb5-2aca-56be-d627-49f35f75c502&l=4725
```
==> azure-arm: + dnf_install 30 1 600 moby-containerd-1.6.26-5.cm2
==> azure-arm: + retries=30
==> azure-arm: + wait_sleep=1
==> azure-arm: + timeout=600
==> azure-arm: + shift
==> azure-arm: + shift
==> azure-arm: + shift
==> azure-arm: ++ seq 1 30
==> azure-arm: + for i in $(seq 1 $retries)
==> azure-arm: + dnf install -y moby-containerd-1.6.26-5.cm2
```

**Comment by cameronmeissner** [comment]
will need a rebase from master

File: vhdbuilder/packer/test/linux-vhd-content-test.sh
Comment (djsly) [inline vhdbuilder/packer/test/linux-vhd-content-test.sh:198]: what are the possible value for `downloadURL` ?

File: parts/linux/cloud-init/artifacts/ubuntu/cse_install_ubuntu.sh
Comment (Devinwong) [inline parts/linux/cloud-init/artifacts/ubuntu/cse_install_ubuntu.sh:111]: The logic of this function is mostly from the original func `installStandaloneContainerd`

File: parts/linux/cloud-init/artifacts/cse_install.sh
Comment (cameronmeissner) [inline parts/linux/cloud-init/artifacts/cse_install.sh:43]: could we factor this out to a somewhat generic `installComponent` function? seems a lot of this logic will be common across all components we install via the components file

File: parts/linux/cloud-init/artifacts/cse_install.sh
Comment (cameronmeissner) [inline parts/linux/cloud-init/artifacts/cse_install.sh:51]: somewhat confused... how is `returnPackageVersions` populating `PackgeVersions`?

File: parts/linux/cloud-init/artifacts/cse_install.sh
Comment (cameronmeissner) [inline parts/linux/cloud-init/artifacts/cse_install.sh:44]: can't we just check `if [ "${OS}" == "$UBUNTU_OS_NAME" ] `?

Comment (cameronmeissner) [reply parts/linux/cloud-init/artifacts/cse_install.sh:44]: ah, I guess we don't really have access to that here, since these functions are supposed to be "common", in that they just work for both ubuntu + mariner (the distro-specific ones are in cse_install_ubuntu.sh/cse_install_mariner.sh)

File: pkg/agent/vhd/cache/types.go
Comment (cameronmeissner) [inline pkg/agent/vhd/cache/types.go:4]: @AlisonB319 do we still need this stuff? I don't think we ever ended up using it for NIC did we?

File: spec/parts/linux/cloud-init/artifacts/test_components.json
Comment (cameronmeissner) [inline spec/parts/linux/cloud-init/artifacts/test_components.json:1]: im assuming we always want this file to reflect the exact state of the actual components.json? right now we'd have to manually keep them in sync right?

might be worth using a symlink instead so we only have to worry about updating a single file

File: spec/parts/linux/cloud-init/artifacts/cse_install_spec.sh
Comment (cameronmeissner) [inline spec/parts/linux/cloud-init/artifacts/cse_install_spec.sh:11]: assuming these are unit tests of some sort, can we add a readme to explain what this is/how they work?

Comment (cameronmeissner) [reply spec/parts/linux/cloud-init/artifacts/cse_install_spec.sh:11]: ah nvm , I see you updated the readme at the root

File: README.md
Comment (cameronmeissner) [inline README.md:84]: could we incorporate installation/usage into the Makefile we already have to keep things simple?

we'd probably want to incorporate this into the `generate` target anyways

File: parts/linux/cloud-init/artifacts/cse_install.sh
Comment (Devinwong) [reply parts/linux/cloud-init/artifacts/cse_install.sh:113]: I didn't have this initially. But without this the ubuntu2204privatekubepkg will fail as it doesn't have containerd definition in components.json. If we are able not to support older image, we can remove this ManifestJson route.

File: parts/linux/cloud-init/artifacts/cse_install.sh
Comment (Devinwong) [reply parts/linux/cloud-init/artifacts/cse_install.sh:51]: I did't find a good way to return an array of strings in Bash scripts. So the func `returnPackageVersions` is actually modifying the global variable `PackageVersions` , defined 1 line above,

File: parts/linux/cloud-init/artifacts/cse_install.sh
Comment (Devinwong) [reply parts/linux/cloud-init/artifacts/cse_install.sh:44]: You are right. We don't know what os it is yet. So we needed to use the command `UBUNTU_RELEASE=$(lsb_release -r -s)` to identify it.

File: parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh
Comment (AlisonB319) [inline parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh:131]: Do we have to have this function because cse_instsall_ubuntu has its own ensureRunc()?

File: pkg/agent/vhd/cache/types.go
Comment (AlisonB319) [reply pkg/agent/vhd/cache/types.go:4]: I think there is leftover stuff from api.GetCachedVersionsOnVHD that I will create a new PR to remove

File: parts/linux/cloud-init/artifacts/cse_install.sh
Comment (Devinwong) [reply parts/linux/cloud-init/artifacts/cse_install.sh:43]: Hmm. There are 2 reasons I didn't do that. But I am open to that.
1. (Minor) This PR has been quite large. Refactoring the download/install logic will introduce another aspect of complexity which will make it even harder to review.
2. (Major) The download/install logic of each package is quite different. I left a comment in the code that
```*)
      echo "Package name: ${name} not supported for download. Please implement the download logic in the script."
      # We can add a common function to download a generic package here.
      # However, installation could be different for different packages.
      ;;
```

For the 5 packages we have now, we are already using various methods to download/install.
1. cri-tools: download using curl and install after artifact streaming. Not sure if it has to install after that.
2. azure-cni: download a tar ball using curl and untar
3. cni-plugins: similar to azure-cni
4. runc: For Ubuntu, it downloads a deb file using curl and install the deb file. For Mariner, it's included in the base image so do nothing.
5. containerd:
- Ubuntu: download/install either through url or apt-get
- Mariner: using dnf to install

We can still have a generic installComponent. But inside it will still check what package it is to use the corresponding method to download/install.

File: parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh
Comment (Devinwong) [reply parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh:131]: In my install-depenedencies.sh around ln 232, if the package is 'runc', ensureRunc will be called. Therefore in both cse_install_ubuntu and cse_install_mariner it needs to have the function. Another way to do that is add an if before that line, if it's Ubuntu, call ensureRunc. Otherwise do nothing (maybe leave the same echo there).

File: README.md
Comment (Devinwong) [reply README.md:84]: Will do! If time does not allow, I will do it in the next PR. (Just trying to unblock other folks with this PR)

File: parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh
Comment (AlisonB319) [reply parts/linux/cloud-init/artifacts/mariner/cse_install_mariner.sh:131]: I'll leave it up to you on what you think is more readable since you're more familiar with this block of code :)

File: spec/parts/linux/cloud-init/artifacts/test_components.json
Comment (Devinwong) [reply spec/parts/linux/cloud-init/artifacts/test_components.json:1]: done!

File: pkg/agent/vhd/cache/types.go
Comment (cameronmeissner) [reply pkg/agent/vhd/cache/types.go:4]: removed in a separate PR

File: parts/linux/cloud-init/artifacts/cse_helpers.sh
Comment (anujmaheshwari1) [inline parts/linux/cloud-init/artifacts/cse_helpers.sh:570]: supernit, comment still has downloadURIs.ubuntu

File: parts/linux/cloud-init/artifacts/cse_helpers.sh
Comment (anujmaheshwari1) [inline parts/linux/cloud-init/artifacts/cse_helpers.sh:552]: this function can be really simple without the if conditionals, correct me if I am wrong:

why we need conditionals is downloadURIs.(ubuntu/mariner/default).${release} right? everything else is the same.

how about we try to access downloadURIs.{osName} and if it doesn't exist, go to downURIs.default?

we can get rid of the osName conditionals then and make the function really simple.

File: parts/linux/cloud-init/artifacts/cse_install.sh
Comment (anujmaheshwari1) [inline parts/linux/cloud-init/artifacts/cse_install.sh:125]: nit: can we preserve the casing for variables? we have upper case everywhere in cse scripts

File: parts/linux/cloud-init/artifacts/cse_helpers.sh
Comment (Devinwong) [reply parts/linux/cloud-init/artifacts/cse_helpers.sh:552]: Good suggestion! Didn't realize the codes were actually duplicated. Updated as you suggested.

File: parts/linux/cloud-init/artifacts/cse_helpers.sh
Comment (Devinwong) [reply parts/linux/cloud-init/artifacts/cse_helpers.sh:570]: Modified the comment to fit the new approach

File: parts/linux/cloud-init/artifacts/cse_helpers.sh
Comment (Devinwong) [reply parts/linux/cloud-init/artifacts/cse_helpers.sh:502]: Good suggestion! Updated as you suggested and add a unit test to cover it.

File: parts/linux/cloud-init/artifacts/cse_install.sh
Comment (Devinwong) [reply parts/linux/cloud-init/artifacts/cse_install.sh:125]: Actually I found multiple naming convention in a single .sh file. For example, `images_to_delete`, `NEEDS_CONTAINERD`, `allMCRImages`. That's why I picked one which is
```
ALL_UPPER_CASE for constants, global variables and environment variables
camelCase for local variables.
```
We can definitely revise and unify all the variables. Does this sound make sense?

File: parts/linux/cloud-init/artifacts/cse_install.sh
Comment (cameronmeissner) [reply parts/linux/cloud-init/artifacts/cse_install.sh:113]: I think we can probably just rebuild the VHD for that particular e2e scenario, that's something we should do anyways

File: parts/linux/cloud-init/artifacts/cse_install.sh
Comment (cameronmeissner) [reply parts/linux/cloud-init/artifacts/cse_install.sh:125]: yeah, Anuj and I were talking about conditional syntax being inconsistent and all over the place as well...

maybe would be better to do that sort of thing in a separate PR

File: parts/linux/cloud-init/artifacts/cse_install.sh
Comment (cameronmeissner) [reply parts/linux/cloud-init/artifacts/cse_install.sh:113]: while we take care of that we can also temporarily disable the scenario

File: parts/linux/cloud-init/artifacts/cse_helpers.sh
Comment (cameronmeissner) [inline parts/linux/cloud-init/artifacts/cse_helpers.sh:506]: are we sure this works properly? seems like it would be assigning to the locally-scoped `release` var

File: parts/linux/cloud-init/artifacts/components.json
Comment (cameronmeissner) [inline parts/linux/cloud-init/artifacts/components.json:46]: why remove 1.6.0?

File: parts/linux/cloud-init/artifacts/components.json
Comment (cameronmeissner) [inline parts/linux/cloud-init/artifacts/components.json:108]: seems like this would also remove 0.2.0, assuming these are typos?

File: parts/linux/cloud-init/artifacts/cse_helpers.sh
Comment (Devinwong) [reply parts/linux/cloud-init/artifacts/cse_helpers.sh:506]: This function will return a string by `echo "${release}"` and the caller will capture it. I have a unit test to cover this so it's working as expected. I run also run a complete abe2e to ensure.

File: parts/linux/cloud-init/artifacts/components.json
Comment (Devinwong) [reply parts/linux/cloud-init/artifacts/components.json:46]: Oh didn't see this. I didn't delete them intentionally. Let me add all back.
Thanks for catching this.

File: parts/linux/cloud-init/artifacts/cse_helpers.sh
Comment (cameronmeissner) [reply parts/linux/cloud-init/artifacts/cse_helpers.sh:506]: so one thing we can do here is assign the output to a non-local variable within the function, then the caller will actually have access to it, that's actually how we do it in our release scripts

I'd personally prefer that over the caller having to capture stdout

File: parts/linux/cloud-init/artifacts/components.json
Comment (Devinwong) [reply parts/linux/cloud-init/artifacts/components.json:398]: fixed!

File: parts/linux/cloud-init/artifacts/cse_helpers.sh
Comment (Devinwong) [reply parts/linux/cloud-init/artifacts/cse_helpers.sh]: Updated.
Though I personally like to echo it so that we don't need so many global variables. But understand that it's the script convention.

**Review by cameronmeissner — CHANGES_REQUESTED** [review]
seems like from the latest abe2e run mariner is trying to install the containerd version for ubuntu, which doesn't exist:

```
++ echo 1.7.15
+ VERSION_B=1.7.15
+ [[ 1.6.26 == \1\.\7\.\1\5 ]]
++ echo 1.6.26 1.7.15
++ sort -V
++ tr ' ' '\n'
+ sorted='1.6.26
1.7.15'
++ IFS=
++ echo '1.6.26
1.7.15'
++ cut '-d
' -f2
+ highestVersion=1.7.15
+ [[ 1.6.26 == 1.7.15 ]]
+ return 1
+ echo 'installing containerd version 1.7.15'
installing containerd version 1.7.15

...

++ seq 1 30
+ for i in $(seq 1 $retries)
+ dnf install -y moby-containerd-1.7.15
environment: line 4: ctr: command not found
Executed "bash -c cleanupImagesRun" 2 times
Fri Jul 12 00:15:24 UTC 2024,abtest7lln000000, endCleanUpKubeProxyImages
Last metadata expiration check: 7:13:34 ago on Thu Jul 11 17:01:47 2024.
No match for argument: moby-containerd-1.7.15
Error: Unable to find a match: moby-containerd-1.7.15
```

**Comment by Devinwong** [comment]
> seems like from the latest abe2e run mariner is trying to install the containerd version for ubuntu, which doesn't exist:
> 
> ```
> ++ echo 1.7.15
> + VERSION_B=1.7.15
> + [[ 1.6.26 == \1\.\7\.\1\5 ]]
> ++ echo 1.6.26 1.7.15
> ++ sort -V
> ++ tr ' ' '\n'
> + sorted='1.6.26
> 1.7.15'
> ++ IFS=
> ++ echo '1.6.26
> 1.7.15'
> ++ cut '-d
> ' -f2
> + highestVersion=1.7.15
> + [[ 1.6.26 == 1.7.15 ]]
> + return 1
> + echo 'installing containerd version 1.7.15'
> installing containerd version 1.7.15
> 
> ...
> 
> ++ seq 1 30
> + for i in $(seq 1 $retries)
> + dnf install -y moby-containerd-1.7.15
> environment: line 4: ctr: command not found
> Executed "bash -c cleanupImagesRun" 2 times
> Fri Jul 12 00:15:24 UTC 2024,abtest7lln000000, endCleanUpKubeProxyImages
> Last metadata expiration check: 7:13:34 ago on Thu Jul 11 17:01:47 2024.
> No match for argument: moby-containerd-1.7.15
> Error: Unable to find a match: moby-containerd-1.7.15
> ```

@cameronmeissner  can you link me to this error? It seems I couldn't find the same error in this e2e logs. https://msazure.visualstudio.com/CloudNativeCompute/_build/results?buildId=97984006&view=logs&j=beeeaba8-323e-5ce9-426a-8cbca9ffe3b0&t=b32dc13d-e9ba-5292-9703-579f43e16486

File: README.md
Comment (Devinwong) [reply README.md:90]: done

File: Makefile
Comment (cameronmeissner) [inline Makefile:220]: nit: `validate-components`?

**Review by cameronmeissner — APPROVED** [review]

**Review by anujmaheshwari1 — APPROVED** [review]

File: Makefile
Comment (Devinwong) [reply Makefile]: updated

File: parts/linux/cloud-init/artifacts/cse_install.sh
Comment (Devinwong) [reply parts/linux/cloud-init/artifacts/cse_install.sh:113]: I will remove installContainerdWithManifestJson once I build a newer ubuntu2204privatekubepkg VHD.

File: spec/parts/linux/cloud-init/artifacts/test_components.json
Comment (cameronmeissner) [inline spec/parts/linux/cloud-init/artifacts/test_components.json:1]: non-blocking, but why do we actually need a separate file at all? can't we just use the original file directly? (which is basically the same as using a symlink)

Comment (Devinwong) [reply spec/parts/linux/cloud-init/artifacts/test_components.json:1]: You are right. Updated again.

**Pull request merged**
